### PR TITLE
Replace JSHint/JSCS with ESLint

### LINKED ISSUE #161: ESLint checking for Ember.js
I started replacing JSHint and JSCS with ESLint in the Ember.js repo in https://github.com/emberjs/ember.js/pull/13549. The next step would be to make it part of the QUnit test suite and I assume that I need to modify this repo for that. Since I'm not that familiar with the whole build process yet it would be awesome to get some pointers on how to do this properly. 

/cc @rwjblue @stefanpenner @krisselden

### PR DESCRIPTION
**Pull request by Turbo87**
This PR enables https://github.com/emberjs/ember.js/pull/13549 to be merged by replacing JSHint and JSCS with ESLint.

Resolves #161 

/cc @rwjblue @stefanpenner

### CODE REVIEW

**Comment by Turbo87** [comment]
this doesn't seem to work properly yet and I'm not quite sure why...

instructions:

- check out the branch from https://github.com/emberjs/ember.js/pull/13549
- install the branch from this PR into it
- produce a lint issue in the `ember.js` repo
- run `ember test --server`
- tests *should* fail, but they seem to succeed

@BrianSipple any ideas?

**Comment by stefanpenner** [comment]
i like were this is going!

File: lib/eslint-tree.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+var path = require('path');
+var ESLintTree = require('broccoli-lint-eslint');
+var escape = require('js-string-escape');
+
+module.exports = function(tree) {
+  var esLintTree = new ESLintTree(tree, {
+    testGenerator: testGenerator,
+  });
+
+  esLintTree.targetExtension = 'eslint.js';
```
Comment (Turbo87) [inline lib/eslint-tree.js]: It looks like this line was causing the issues. If the default `lint-test.js` extension is used instead it works properly...

**Comment by Turbo87** [comment]
@rwjblue @stefanpenner this should be ready for merging now. you can test this with the instructions above.

**Comment by Turbo87** [comment]
TravisCI failures seem unrelated to this PR

**Comment by stefanpenner** [comment]
I'll try to find time over the next few days, feel free to bug me if i dont.

**Comment by Turbo87** [comment]
@stefanpenner ping ;)

**Comment by Turbo87** [comment]
🐛

**Comment by stefanpenner** [comment]
👀

**Comment by stefanpenner** [comment]
running this i get

```
File: ember-metal/tests/features_test.js
The Broccoli Plugin: [Babel] failed with:
Error: ember-metal/tests/features_test.js: [BABEL] ember-metal/tests/features_test.js: An unknown feature 'fred'' was encountered
    at Logger.error (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/transformation/file/logger.js:58:11)
    at NodePath.babel.Transformer.CallExpression (/Users/stefanpenner/src/ember.js/node_modules/babel-plugin-feature-flags/index.js:42:22)
    at NodePath.call (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/path/context.js:56:28)
    at NodePath.visit (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/path/context.js:90:8)
    at TraversalContext.visitMultiple (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/context.js:108:16)
    at TraversalContext.visit (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/context.js:146:19)
    at Function.traverse.node (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/index.js:76:17)
    at NodePath.visit (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/path/context.js:107:26)
    at TraversalContext.visitSingle (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/context.js:132:12)
    at TraversalContext.visit (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/context.js:148:19)
    at Function.traverse.node (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/index.js:76:17)
    at NodePath.visit (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/path/context.js:107:26)
    at TraversalContext.visitMultiple (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/context.js:108:16)
    at TraversalContext.visit (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/context.js:146:19)
    at Function.traverse.node (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/index.js:76:17)
    at NodePath.visit (/Users/stefanpenner/src/ember.js/node_modules/babel-core/lib/traversal/path/context.js:107:26)

```

which may be unrelated?

**Comment by stefanpenner** [comment]
commenting out the test above ^, i get further. It seems like i still see

```
===== 1 JSHint Error
```

in the console is that expected?

**Comment by rwjblue** [comment]
@stefanpenner - That test has been removed from ember master, because it was fundamentally flawed (and an update to babel-plugin-feature-flags broke it).

**Comment by stefanpenner** [comment]
@rwjblue thanks

**Comment by stefanpenner** [comment]
@Turbo87 once i actually read your instructions and switched to the correct emberjs-build branch i seem to get:

```
The Broccoli Plugin: [Creator: ember/version.js] failed with:
Error: ENOENT: no such file or directory, symlink '/Users/stefanpenner/src/ember.js/tmp/creator-cache_path-NJJm6B40.tmp/ember/version.js' -> '/Users/stefanpenner/src/ember.js/tmp/
creator-output_path-vMMfS5sw.tmp/ember/version.js'
    at Error (native)
    at Object.fs.symlinkSync (fs.js:899:18)
    at symlink (/Users/stefanpenner/src/emberjs-build/node_modules/broccoli-file-creator/node_modules/symlink-or-copy/index.js:82:14)
    at symlinkOrCopySync (/Users/stefanpenner/src/emberjs-build/node_modules/broccoli-file-creator/node_modules/symlink-or-copy/index.js:58:5)
    at linkFromCache (/Users/stefanpenner/src/emberjs-build/node_modules/broccoli-file-creator/index.js:43:3)
    at Creator.build (/Users/stefanpenner/src/emberjs-build/node_modules/broccoli-file-creator/index.js:31:3)
    at /Users/stefanpenner/src/emberjs-build/node_modules/broccoli-file-creator/node_modules/broccoli-plugin/read_compat.js:61:34
    at lib$rsvp$$internal$$tryCatch (/Users/stefanpenner/src/emberjs-build/node_modules/broccoli-file-creator/node_modules/rsvp/dist/rsvp.js:493:16)
    at lib$rsvp$$internal$$invokeCallback (/Users/stefanpenner/src/emberjs-build/node_modules/broccoli-file-creator/node_modules/rsvp/dist/rsvp.js:505:17)
    at /Users/stefanpenner/src/emberjs-build/node_modules/broccoli-file-creator/node_modules/rsvp/dist/rsvp.js:1001:13
    at lib$rsvp$asap$$flush (/Users/stefanpenner/src/emberjs-build/node_modules/broccoli-file-creator/node_modules/rsvp/dist/rsvp.js:1198:9)
    at doNTCallback0 (node.js:408:9)
    at process._tickCallback (node.js:337:13)

The broccoli plugin was instantiated at:
    at Creator.Plugin (/Users/stefanpenner/src/emberjs-build/node_modules/broccoli-file-creator/node_modules/broccoli-plugin/index.js:7:31)
    at new Creator (/Users/stefanpenner/src/emberjs-build/node_modules/broccoli-file-creator/index.js:15:10)
    at createEmberVersion (/Users/stefanpenner/src/emberjs-build/lib/create-ember-version.js:8:14)
    at CoreObject.enumeratePackages [as _enumeratePackages] (/Users/stefanpenner/src/emberjs-build/lib/ember-build.js:204:45)
    at CoreObject.generateCompiledSourceTree [as _generateCompiledSourceTree] (/Users/stefanpenner/src/emberjs-build/lib/ember-build.js:95:26)
    at CoreObject.getDistTrees (/Users/stefanpenner/src/emberjs-build/lib/ember-build.js:279:34)
    at module.exports (/Users/stefanpenner/src/ember.js/ember-cli-build.js:97:21)
    at CoreObject.module.exports.Task.extend.setupBroccoliBuilder (/Users/stefanpenner/src/ember.js/node_modules/ember-cli/lib/models/builder.js:54:19)
    at CoreObject.module.exports.Task.extend.init (/Users/stefanpenner/src/ember.js/node_modules/ember-cli/lib/models/builder.js:89:10)
    at CoreObject.superWrapper [as init] (/Users/stefanpenner/src/ember.js/node_modules/core-object/lib/assign-properties.js:32:18)
    at CoreObject.Class (/Users/stefanpenner/src/ember.js/node_modules/core-object/core-object.js:32:33)
    at CoreObject.module.exports.Task.extend.run (/Users/stefanpenner/src/ember.js/node_modules/ember-cli/lib/tasks/serve.js:15:19)
    at /Users/stefanpenner/src/ember.js/node_modules/ember-cli/lib/commands/serve.js:64:24
    at lib$rsvp$$internal$$tryCatch (/Users/stefanpenner/src/ember.js/node_modules/rsvp/dist/rsvp.js:1036:16)
    at lib$rsvp$$internal$$invokeCallback (/Users/stefanpenner/src/ember.js/node_modules/rsvp/dist/rsvp.js:1048:17)
    at /Users/stefanpenner/src/ember.js/node_modules/rsvp/dist/rsvp.js:331:11
```

thoughts?

**Comment by Turbo87** [comment]
@stefanpenner hmm that looks unrelated. did you make sure to clean all the caches?

**Comment by Turbo87** [comment]
@stefanpenner can't reproduce your second issue. I'm getting hit by `An unknown feature 'fred'' was encountered` too, but that should be solved once https://github.com/emberjs/ember.js/pull/13549 is rebased.

**Comment by stefanpenner** [comment]
As discussed in slack, we will need to adjust how the eslint plugin is loaded, so that `npm link` of emberjs-build works.

I wasn't able to fully test,  (headless tests, appear unhappy on my machine but unrelated to this PR), but the rest seems to work good and as expected.

**Comment by Turbo87** [comment]
> As discussed in slack, we will need to adjust how the eslint plugin is loaded, so that npm link of emberjs-build works.

unfortunately this doesn't seem to be possible currently. the only way to specify plugins right now is through their package name (`eslint-plugin-ember-internal` or just `ember-internal`) which will be resolved relative to the `eslint` installation. if `npm link` is used for `emberjs-build` those plugins will not be in a parent path of `emberjs-build` and so they won't be discovered. the only way that this might work with `npm link` is if the plugins are installed in `emberjs-build` too.

this might be off-topic but what was the reason for extracting `emberjs-build` out of the `ember.js` repo? I guess if the build pipeline was part of `ember.js` too this problem wouldn't actually exist.

**Comment by stefanpenner** [comment]
> this might be off-topic but what was the reason for extracting emberjs-build out of the ember.js repo? I guess if the build pipeline was part of ember.js too this problem wouldn't actually exist.

I'm not sure the abstractions where quite correct, but this is the current state of things.  I do believe there is common sentiment to dramatically improve this.

We should chat sometime this week, if eslint cannot be abstracted like this we may need to work to solve it.

**Comment by Turbo87** [comment]
> We should chat sometime this week, if eslint cannot be abstracted like this we may need to work to solve it.

The issue with that is that ESLint 3 was just released a few days ago which dropped Node 0.x support. My guess is that ESLint 2 won't receive any new features or refactorings in that direction so we would have to either drop Node 0.x support for Ember development or live without npm linking this in. (or fork ESLint, but I don't think that would be a proper long term solution)

**Comment by stefanpenner** [comment]
> (or fork ESLint, but I don't think that would be a proper long term solution)

:s, sounds like we need to add this to the agenda

**Comment by Turbo87** [comment]
I'm wondering if @nzakas would answer if I asked if this was something that he would consider for another v2 release...

for context: we are talking about using `require('some-package')` in `plugins` in `.eslintrc.js`

**Comment by Turbo87** [comment]
and if we'd really fork for the time being then we'd also have to fork `broccoli-lint-eslint`... :-/

**Comment by rwjblue** [comment]
FWIW, I think we can require a specific node version to build Ember itself...

**Comment by stefanpenner** [comment]
> FWIW, I think we can require a specific node version to build Ember itself...

I don't believe so, as we would like people to be able to build ember like they do ember-data. That being said, we could disable eslint in those cases.

**Comment by rwjblue** [comment]
Agreed. Consumers of addons typically do not run linting or tests for those addons.

**Comment by Turbo87** [comment]
@stefanpenner it seems that https://github.com/eslint/eslint/pull/6865 might solve our issue

**Comment by Turbo87** [comment]
@stefanpenner @rwjblue this PR is up-to-date again and IMHO ready to be merged.

instructions to try it out:

- check out the branch from emberjs/ember.js#13549
- `npm link emberjs-build`
- `ember test --server`
- produce a lint issue in the ember.js repo (e.g. remove a `@public` tag from the docs)

**Comment by Turbo87** [comment]
@stefanpenner FYI I worked around the plugin resolution issue by just not using any plugins. It seems that having custom rules in the repository is also well supported by ESLint and since we can give absolute paths to the folder with the custom rules there is no issue with `npm link` either.

File: lib/ember-build.js [hunk 2]
```diff
@@ -233,7 +233,8 @@ var EmberBuild = CoreObject.extend({
       currentPackage.trees = es6Package(packages, packageName, {
         htmlbars: this.htmlbars,
         glimmer: this.glimmer,
-        vendoredPackages: vendoredPackages
+        vendoredPackages: vendoredPackages,
+        eslintRulePaths: ['lib/eslint-rules'],
```
Comment (rwjblue) [inline lib/ember-build.js]: This should be passed in from ember-cli-build.js in Ember itself.

File: package.json [hunk 3]
```diff
@@ -39,9 +39,8 @@
     "broccoli-concat": "^2.0.0",
     "broccoli-file-creator": "^1.0.0",
     "broccoli-funnel": "^1.0.1",
-    "broccoli-jscs": "^1.4.1",
-    "broccoli-jshint": "^1.1.0",
     "broccoli-kitchen-sink-helpers": "^0.2.6",
+    "broccoli-lint-eslint": "^2.3.0",
```
Comment (rwjblue) [inline package.json]: 3.x

**Comment by rwjblue** [comment]
A few small changes here, and I think this can land...

File: package.json (hunk 3 above)
Comment (Turbo87) [inline package.json]: yeah, that makes more sense now 👍

**Comment by Turbo87** [comment]
@rwjblue done

**Comment by rwjblue** [comment]
Looks like Travis is failing due to 0.12 / 0.10 now, I submitted https://github.com/emberjs/emberjs-build/pull/185 to remove those and will re-kick off the build here (not sure if it will fix it without a rebase though).

**Comment by Turbo87** [comment]
@rwjblue rebased, thanks!

**Comment by rwjblue** [comment]
Thank **you**!

**Comment by rwjblue** [comment]
Released as 0.19.0

**Pull request merged**